xayn-ai/src/embedding/utils.rs

```diff
@@ -1,90 +1,166 @@
+/// Computes the l2 norm (euclidean metric) of a vector.
 ///
 /// # Panics
-/// Panics if the vectors didn't consist of all real values.
-pub fn l2_distance<A, B, S>(a: &A, b: &B) -> f32
+/// Panics if the vector doesn't consist solely of real values.
+pub fn l2_norm<S>(a: ArrayBase<S, Ix1>) -> f32
 where
-    A: Deref<Target = ArrayBase<S, Ix1>>,
-    B: Deref<Target = ArrayBase<S, Ix1>>,
     S: Data<Elem = f32>,
 {
-    let difference = a.deref() - b.deref();
-    let distance = difference.dot(&difference).sqrt();
-
-    if distance.is_nan() || distance.is_infinite() {
-        panic!(
-            "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
-            a.deref(),
-            b.deref(),
-        );
-    }
+    let norm = a.dot(&a).sqrt();
+    assert!(
+        norm.is_finite(),
+        "vector must consist of real values only, but got:\n{:?}",
+        a,
+    );
+
+    norm
+}
 
-    distance
+/// Computes the l2 norm (euclidean metric) of the difference of two vectors.
+///
+/// # Panics
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+pub fn l2_distance<S>(a: ArrayBase<S, Ix1>, b: ArrayBase<S, Ix1>) -> f32
+where
+    S: Data<Elem = f32>,
+{
+    l2_norm(&a - &b)
 }
 
 /// Computes the arithmetic mean of two vectors.
 ///
 /// # Panics
-/// Panics if the vectors do not consist solely of real values.
-pub fn mean<A, S>(a: &A, b: &A) -> Embedding
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+pub fn mean<S>(a: ArrayBase<S, Ix1>, b: ArrayBase<S, Ix1>) -> Array1<f32>
 where
-    A: Deref<Target = ArrayBase<S, Ix1>>,
     S: Data<Elem = f32>,
 {
-    let mean = 0.5 * (a.deref() + b.deref());
-    if mean.iter().any(|elt| elt.is_nan() || elt.is_infinite()) {
-        panic!(
-            "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
-            a.deref(),
-            b.deref(),
-        );
+    let mean = 0.5 * (&a + &b);
+    assert!(
+        mean.iter().copied().all(f32::is_finite),
+        "vectors must consist of real values only, but got\na: {:?}\nb: {:?}",
+        a,
+        b,
+    );
+
+    mean
+}
+
+/// Computes the pairwise cosine similarities of vectors.
+///
+/// # Panics
+/// Panics if the vectors don't consist solely of real values or their shapes don't match.
+#[allow(dead_code)]
+pub fn pairwise_cosine_similarity<S>(a: &[ArrayBase<S, Ix1>]) -> Array2<f32>
+where
+    S: Data<Elem = f32>,
+{
+    let norms = a.iter().map(|a| l2_norm(a.view())).collect::<Vec<_>>();
+    let size = a.len();
+    let mut similarities = Array2::ones((size, size));
+    for i in 0..size {
+        if norms[i] != 0. {
+            for j in i + 1..size {
+                if norms[j] != 0. {
+                    similarities[[i, j]] = a[i].dot(&a[j]) / norms[i] / norms[j];
+                    similarities[[j, i]] = similarities[[i, j]];
+                }
+            }
+        }
```

Comment:
theoretically the else branch is mathematically undefined i.e. `NaN` is it ok to just set it to `1` for us? (I guess probably, but we maybe should document it).

Reply:
if a norm of a vector is zero, then all its elements have to be zero as well. since the zero vector is pointing in all direction at the same time, it's ok to define the similarity between a vector and the zero vector as 1.

in the end we will not encounter this in the real world, because the compared embeddings are almost never zero vectors.

Reply:
> because the compared embeddings are almost never zero vectors.

ok, so we keep it that way

> since the zero vector is pointing in all direction at the same time,

Or no direction (it's e.g. not uncommon I think to define the 0-vector as orthogonal with all vectors, in which case it would default to 0).  I'm pretty sure it's mathematically undefined (in the sense that there is no agreement on that and it could be either or even neither) and the similarity is normally also not defined for `0` vectors. We can define it the way it is in the code, but then we should document it.

Reply:
fair point. i updated the docs accordingly.

in our case i prefer to define the similarity as 1 to keep it aligned with the pink code (although they are doing some improper computations, for the sake of avoiding division by zero, which results in the similarity to a zero vector to be 10^6  instead of 1).